### PR TITLE
Add private `ToolkitRegistry` context manager

### PR DESCRIPTION
**Pull request by mattwthompson**
Relevant to https://github.com/openforcefield/openff-toolkit/issues/966 and some other discussions.

This uses some of @SimonBoothroyd's code from [`openff-fragmenter`](https://github.com/openforcefield/openff-fragmenter/blob/4a290b866a8ed43eabcbd3231c62b01f0c6d7df6/openff/fragmenter/utils.py#L97-L123) to enable registry selection without needing to implement it in Interchange.

- [x] Tag issue being addressed
- [x] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/master/openff/toolkit/tests)
- [ ] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/master/docs), if applicable - AFAIR this change only adds a private function
- [x] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [ ] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/master/docs/releasehistory.rst)

### CODE REVIEW

**Review by j-wags — COMMENTED** [review]
This could use some tests, for example:

* `create_openmm_system` with OpenEye vs with AT/RDK - Expect different charges
* `create_openmm_system` with AT/RDK, asking for AM1BCCELF10, ensuring that we get an error
* Pass an invalid registry (like an empty one, or only RDKit) and expect an error from charge assignment

**Comment by mattwthompson** [comment]
This isn't quite working the way I hoped in that `GLOBAL_TOOLKIT_REGISTRY` is not restored to its original state when an exception is raise inside the context of `_toolkit_registry_manager`.


```python3
In [1]: from openff.toolkit import *

In [2]: from openff.toolkit.utils.toolkit_registry import _toolkit_registry_manager

In [3]: from openff.toolkit.utils import GLOBAL_TOOLKIT_REGISTRY, BuiltInToolkitWrapper

In [4]: topology = Molecule.from_smiles("CCOCCCOCCOC").to_topology()

In [5]: sage = ForceField("openff-2.0.0.offxml")

In [6]: sage.create_openmm_system(topology, toolkit_registry=BuiltInToolkitWrapper(), use_interchange=True)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 sage.create_openmm_system(topology, toolkit_registry=BuiltInToolkitWrapper(), use_interchange=True)

<SNIPPING SOME VERBOSE TRACEBACK>

File ~/software/openff-toolkit/openff/toolkit/utils/toolkit_registry.py:382, in ToolkitRegistry.call(self, method_name, raise_exception_types, *args, **kwargs)
    380 for toolkit, error in errors:
    381     msg += " {} {} : {}\n".format(toolkit, type(error), error)
--> 382 raise ValueError(msg)

ValueError: No registered toolkits can provide the capability "find_smarts_matches" for args "(Molecule with name '' with bad SMILES and Hill formula 'C8H18O3', '[#6X4:1]-[#6X4:2]')" and kwargs "{'unique': False}"
Available toolkits are: [ToolkitWrapper around Built-in Toolkit version None]


In [7]: GLOBAL_TOOLKIT_REGISTRY
Out[7]: ToolkitRegistry containing Built-in Toolkit
```

Everything is as expected until the last line - I thought the point of these lines

https://github.com/openforcefield/openff-toolkit/blob/9b4f0e05671fca4da6fc86b416ef9535bf5dfbc6/openff/toolkit/utils/toolkit_registry.py#L412-L418

was to undo the modifications that happened a few lines above.

**Comment by j-wags** [comment]
Oof, you're right. I looked into it and context managers don't handle exceptions gracefully by default. 

So, one option to move forward would be to not use the contextlib approach and instead make our own contextmanager class with a custom `__enter__` and `__exit__` method.

Or, the other option, which I tried implementing in that last commit, is to check if there's an exception in the `yield` block, and then reset the GLOBAL_TOOLKIT_REGISTRY and raise the exception once everything's back to normal. 

Apologies that I was wrong about this initially. I think the manually-raise-the-exception approach should be fine if that works for you.

File: openff/toolkit/utils/toolkit_registry.py [hunk 1]
```diff
@@ -383,3 +385,41 @@ def __repr__(self):
...
+        exc = e
+
+    for toolkit in context_toolkits:
+        GLOBAL_TOOLKIT_REGISTRY.deregister_toolkit(toolkit)
+
+    for toolkit in original_toolkits:
+        GLOBAL_TOOLKIT_REGISTRY.register_toolkit(toolkit)
+
+    if exc is not None:
+        raise exc
```
Comment (SimonBoothroyd) [inline openff/toolkit/utils/toolkit_registry.py:425]:
```suggestion
    try:
        yield
    finally:
        for toolkit in context_toolkits:
            GLOBAL_TOOLKIT_REGISTRY.deregister_toolkit(toolkit)
        for toolkit in original_toolkits:
            GLOBAL_TOOLKIT_REGISTRY.register_toolkit(toolkit)
```

**Comment by SimonBoothroyd** [comment]
Ah you're right I missed the finally statement! See e.g. [here](https://github.com/openforcefield/openff-utilities/blob/f9f623823b4c49162bdd68ad88c8bf0469063036/openff/utilities/utilities.py#L150) were we correctly handle potential errors in a context manager like this

**Comment by mattwthompson** [comment]
Thanks @SimonBoothroyd!

**Comment by mattwthompson** [comment]
Okay, I think this is ready to go now.

**Review by j-wags — APPROVED** [review]
Looks great!

**Pull request merged**